report/topology.go

```diff
@@ -311,6 +331,119 @@ func (s StringSet) Copy() StringSet {
 	return result
 }
 
+// Metrics is a string->metric map.
+type Metrics map[string]Metric
+
+// Merge merges two sets maps into a fresh set, performing set-union merges as
+// appropriate.
+func (m Metrics) Merge(other Metrics) Metrics {
+	result := m.Copy()
+	for k, v := range other {
+		result[k] = result[k].Merge(v)
+	}
+	return result
+}
+
+// Copy returns a value copy of the sets map.
+func (m Metrics) Copy() Metrics {
+	result := Metrics{}
+	for k, v := range m {
+		result[k] = v.Copy()
+	}
+	return result
+}
+
+// Metric is a list of timeseries data. Clients must use the Add
+// method to add values.
+type Metric struct {
+	Samples []Sample  `json:"samples"`
+	Min     float64   `json:"min"`
+	Max     float64   `json:"max"`
+	First   time.Time `json:"first"`
+	Last    time.Time `json:"last"`
+}
+
+type Sample struct {
+	Timestamp time.Time `json:"date"`
+	Value     float64   `json:"value"`
+}
+
+// MakeMetric makes a new Metric.
+func MakeMetric() Metric {
+	return Metric{}
+}
+
+func (m Metric) WithFirst(t time.Time) Metric {
+	m.First = t
+	return m
+}
+
+// Add adds the sample to the Metric. Add is the only valid way to grow a
+// Metric. Add returns the Metric to enable chaining.
+func (m Metric) Add(t time.Time, v float64) Metric {
+	i := sort.Search(len(m.Samples), func(i int) bool { return !t.After(m.Samples[i].Timestamp) })
+	if i < len(m.Samples) && m.Samples[i].Timestamp.Equal(t) {
+		// The list already has the element.
+		return m
+	}
+	// It is a new element, insert it in order.
+	m.Samples = append(m.Samples, Sample{})
+	copy(m.Samples[i+1:], m.Samples[i:])
+	m.Samples[i] = Sample{Timestamp: t, Value: v}
+	if v > m.Max {
+		m.Max = v
+	}
+	if v < m.Min {
+		m.Min = v
+	}
+	if m.First.IsZero() || t.Before(m.First) {
+		m.First = t
+	}
+	if m.Last.IsZero() || t.After(m.Last) {
+		m.Last = t
+	}
+	return m
+}
+
+// Merge combines the two Metrics and returns a new result.
+func (m Metric) Merge(other Metric) Metric {
```

Comment:
No need for a copy, as everything in it is immutable, and `m` is passed-by-value.